include/fir/Type.h

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FIR_TYPE_H
+#define FIR_TYPE_H
+
+#include "mlir/IR/Types.h"
+#include <variant>
+
+namespace llvm {
+class StringRef;
+template<typename> class ArrayRef;
+class hash_code;
+}
+
+namespace fir {
+
+class FIROpsDialect;
+
+using KindTy = int;
+
+namespace detail {
+struct FIRBoxTypeStorage;
+struct FIRBoxCharTypeStorage;
+struct FIRBoxProcTypeStorage;
+struct FIRCharacterTypeStorage;
+struct FIRCplxTypeStorage;
+struct FIRDimsTypeStorage;
+struct FIRFieldTypeStorage;
+struct FIRHeapTypeStorage;
+struct FIRIntTypeStorage;
+struct FIRLogicalTypeStorage;
+struct FIRPointerTypeStorage;
+struct FIRRealTypeStorage;
+struct FIRRecordTypeStorage;
+struct FIRReferenceTypeStorage;
+struct FIRSequenceTypeStorage;
+struct FIRTypeDescTypeStorage;
+}
+
+enum FIRTypeKind {
+  // The enum starts at the range reserved for this dialect.
+  FIR_TYPE = mlir::Type::FIRST_FIR_TYPE,
+  FIR_BOX,
+  FIR_BOXCHAR,
+  FIR_BOXPROC,
+  FIR_CHARACTER,  // intrinsic
+  FIR_COMPLEX,  // intrinsic
+  FIR_DERIVED,  // derived
+  FIR_DIMS,
+  FIR_FIELD,
+  FIR_HEAP,
+  FIR_INT,  // intrinsic
+  FIR_LOGICAL,  // intrinsic
+  FIR_POINTER,
+  FIR_REAL,  // intrinsic
+  FIR_REFERENCE,
+  FIR_SEQUENCE,
+  FIR_TYPEDESC,
+};
+
+template<class... Ts> struct Typecase : Ts... { using Ts::operator()...; };
+template<class... Ts> Typecase(Ts...)->Typecase<Ts...>;
+
+bool isa_fir_type(mlir::Type);
+bool isa_std_type(mlir::Type t);
+bool isa_fir_or_std_type(mlir::Type t);
+
+// Intrinsic types
+
+class CharacterType : public mlir::Type::TypeBase<CharacterType, mlir::Type,
```

Comment:
For `CharacterType`,  `IntType`, `LogicalType`, `LogicalType` and `CplxType`, there may be an opportunity to define a base class:
```
template<typename T, FIRTypeKind K>
class IntrinsicType {
public:
  static T get(mlir::MLIRContext *ctxt, KindTy kind);
  static bool kindof(unsigned kind) {return kind == K;}
  int getSizeInBits() const;
  KindTy getFKind() const;
};
```
and have  `CharacterType` be:
```
class CharacterType : public
           mlir::Type::TypeBase<CharacterType, mlir::Type, detail::FIRCharacterTypeStorage>,
           IntrinsicType<CharacterType, FIRTypeKind::FIR_CHARACTER> {
public:
 using Base::Base;
};
```
IMHO, it makes it clearer that they have the same interface (maybe it could be problematic for the implementation part though, I do not know). Maybe you can also use something similar for classes below, but it seems harder to share things but for `getEleTy` and `kindof`.